experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -0,0 +1,714 @@
+    R = parent(v[1])
+    N = modulus(R)
+    A = rand(Int, 1000)
+    for a in A
+        z = R(a)^3 + R(a) * v[1] + v[2]
+        if jacobi_symbol(ZZ(z), N) != -1
+            (bool, res) = issquare_with_square_root(z)
+            if bool
+                return (bool, [R(a), res, R(1)])
+            end
+        end
+    end
+    return (false, R.([0, 0, 0]))
+end
+
+################################################################################
+
+function compute_p1p2(v::Vector{fmpz_mod}, m::fmpz, q::fmpz)
+    (bool, P) = random_point(v)
+    R = parent(v[1])
+    !bool && return (false, (R.([0, 0, 0]), ZZ(1)), (R.([0, 0, 0]), ZZ(1)))
+    P2 = _scalar_mult(P, v, div(m, q))
+    P1 = _scalar_mult(P2[1], v, div(m, q))
+    return (true, P1, P2)
+end
+
+################################################################################
+
+function quadratic_non_residue(N::fmpz, D::fmpz)
+    R = ResidueRing(ZZ, N)
+    A = rand(Int, 1000)
+    for a in A
+        (bool, res) = issquare_with_square_root(R(a))
+        if !bool
+            if D == ZZ(-3)
+                !isone(R(a)^div(N - 1, 3)) && return R(a)
+            else
+                return R(a)
+            end
+        end
+    end
+    return R(0)
+end
+
+################################################################################
+
+function atkin_morain_step(
+    N::fmpz,
+    D::fmpz,
+    m::fmpz,
+    q::fmpz,
+    res::Vector{Tuple{fmpz,Int}} = Vector{Tuple{fmpz,Int}}(),
+    pos::Int = 1,
+)
+    (v, h) = compute_ell_curve(N, D)
+    R = parent(v[1])
+    O_E = R.([0, 1, 0])
+    g = quadratic_non_residue(N, D)
+    k = 0
+
+    while k < h
+        P1 = (O_E, ZZ(1))
+        while P1[1] == O_E
+            (bool2, P1, P2) = compute_p1p2(v, m, q)
+            !bool2 && return (false, N, ZZ(1))
+            P2[1] == R.([0, 0, 0]) && return (false, N, P2[2])
+            P1[1] == R.([0, 0, 0]) && return (false, N, P1[2])
+            P2[1] != O_E && return (true, q, ZZ(1), push!(res, (q, pos)), pos)
+        end
+        k = k + 1
+        if D == ZZ(-3)
+            v = [v[1], v[2] * g]
+        elseif D == ZZ(-4)
+            v = [v[1] * g, v[2]]
+        else
+            v = [v[1] * g^2, v[2] * g^3]
+        end
+    end
+    return (false, N, ZZ(1), res, pos)
+end
+
+################################################################################
+
+function atkin_morain(
+    N::fmpz,
+    Arr::Vector{fmpz} = Vector{fmpz}(),
+    pos::Int = 1,
+    res::Vector{Tuple{fmpz,Int}} = Vector{Tuple{fmpz,Int}}(),
+)
+    N < ZZ(10)^15 && return isprime(N)
+    if length(Arr) == 0
+        L = funddiscriminant(10^5)
+    else
+        L = Arr
+    end
+    pos > length(L) && error("No proof found")
+    i = pos
+    while i <= length(L)
+        if jacobi_symbol(L[i], N) == 1
+            (bool1, (m, q)) = find_q(N, L[i])
+            if bool1
+                (bool2, A, B, res, pos) = atkin_morain_step(N, L[i], m, q, res, i)
+                B > 1 && false
```

Comment:
What is this line doing?

Reply:
This is just a leftover from a previous version which I forgot to remove.